core/LSU.cpp

```diff
@@ -216,22 +252,132 @@ namespace olympia
         // NOTE:
         // win_ptr should always point to an instruction ready to be issued
         // Otherwise assertion error should already be fired in arbitrateInstIssue_()
+        ILOG("Arbitrated inst " << win_ptr << " " << win_ptr->getInstPtr());
 
         lsu_insts_issued_++;
 
         // Append load/store pipe
-        ldst_pipeline_.append(win_ptr->getMemoryAccessInfoPtr());
+        ldst_pipeline_.append(win_ptr);
+
+        // We append to replay queue to prevent ref count of the shared pointer to drop before
+        // calling pop below
+        if(allow_speculative_load_exec_)
+        {
+            ILOG("Appending to replay queue " << win_ptr);
+            appendToReplayQueue_(win_ptr);
+        }
+
+        // Remove inst from ready queue
+        ready_queue_.pop();
+        win_ptr->setInReadyQueue(false);
```

Comment:
This is the valgrind error I found.  After the `pop` operation, `win_ptr` is no longer valid (although the memory is still on the stack).  

What is happening:
1. You call `arbitrateInstIssue` which returns the Ptr by reference and you make a `const` reference to be used locally
2. Then, here, you `pop` it and now that pointer is nullified.

Suggestions on fixing:
Have `arbitrateInstIssue` return the pointer by value AND perform the `pop` _in that function_.

Reply:
Looks good.  After making another fix to DCache.hpp, valgrind is clean.  Good stuff!